File: src/model/session.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use tokio_postgres::Row;
use uuid::Uuid;

use crate::db::FromRow;

#[derive(Deserialize, Serialize)]
pub struct Session {
  pub session_id: Uuid,
  pub user_id: Uuid,
  pub app_id: Uuid,
  pub created_at: DateTime<Utc>,
  pub updated_at: DateTime<Utc>,
  pub access_expires_at: DateTime<Utc>,
  pub refresh_expires_at: DateTime<Utc>,
}

impl FromRow for Session {
  fn from_row(row: Row) -> Option<Self> {
    Some(Session {
      session_id: row.get("session_id"),
      user_id: row.get("user_id"),
      app_id: row.get("app_id"),
      created_at: row.get("created_at"),
      updated_at: row.get("updated_at"),
      access_expires_at: row.get("access_expires_at"),
      refresh_expires_at: row.get("refresh_expires_at"),
    })
  }
}

File: ui-ios/Chameleon/Data/Session.swift
import Foundation
import KeychainSwift

struct Session: Codable {
  let id: UUID
  let serverUrl: URL
  let serverName: String
  let serverApiUrl: URL
  let serverCdnUrl: URL
  let sessionId: UUID
  let userId: UUID
  let appId: UUID
  let createdAt: Date
  let updatedAt: Date
  let accessExpiresAt: Date
  let refreshExpiresAt: Date

  func save() async -> Result<Void> {
    do {
      let encoder = JSONEncoder()
      let data = try encoder.encode(self)
      guard let body = String(data: data, encoding: .utf8) else {
        throw CoreError.stringEncode
      }

      let keychain = KeychainSwift(keyPrefix: "uk.lyptt.chameleon")
      keychain.set(body, forKey: id.uuidString)

      UserDefaults.standard.set(id.uuidString, forKey: "last-session")

      return .ok(data: ())
    } catch {
      if let err = error as? CoreError {
        return .error(err: err)
      }

      return .error(err: .unknown)
    }
  }

  static func load() async -> Result<Session> {
    guard let lastSession = UserDefaults.standard.string(forKey: "last-session") else {
      return .error(err: .missingData)
    }

    return await load(id: lastSession)
  }

  static func load(id: String) async -> Result<Session> {
    do {
      let keychain = KeychainSwift(keyPrefix: "uk.lyptt.chameleon")
      let decoder = JSONDecoder()

      guard let body = keychain.get(id) else {
        throw CoreError.missingData
      }

      guard let data = body.data(using: .utf8) else {
        throw CoreError.stringParse
      }

      guard let ret = try? decoder.decode(Session.self, from: data) else {
        throw CoreError.jsonParse
      }

      return .ok(data: ret)
    } catch {
      if let err = error as? CoreError {
        return .error(err: err)
      }

      return .error(err: .unknown)
    }
  }
}
